<!-- Launches Statemachine, Navigation and Explore Lite for SchrÃ¶di -->

<launch>
	<arg name="robot_frame" default="base_footprint" />
	<arg name="autonomy_cmd_vel_topic" default="/autonomy/vel/cmd" />
	<arg name="teleoperation_cmd_vel_topic" default="/vel/teleop" />
	<arg name="cmd_vel_topic" default="vel/autonomy" />
	<arg name="goal_plugin" default="statemachine::CalculateGoalState" />
	<arg name="navigation_plugin"
		default="statemachine::NavigationState" />
	<arg name="mapping_plugin" default="statemachine::MappingState" />
	<arg name="waypoint_routines" default="['Reversing']" />

	<include file="$(find statemachine)/launch/statemachine.launch">
		<arg name="update_frequency" value="20" />
		<arg name="robot_frame" value="$(arg robot_frame)" />
		<arg name="autonomy_cmd_vel_topic"
			value="$(arg autonomy_cmd_vel_topic)" />
		<arg name="teleoperation_cmd_vel_topic"
			value="$(arg teleoperation_cmd_vel_topic)" />
		<arg name="cmd_vel_topic" value="$(arg cmd_vel_topic)" />
		<arg name="calculate_goal_plugin" value="$(arg goal_plugin)" />
		<arg name="navigation_plugin" value="$(arg navigation_plugin)" />
		<arg name="mapping_plugin" value="$(arg mapping_plugin)" />
		<arg name="waypoint_routines" value="$(arg waypoint_routines)" />
		<arg name="exploration_goal_tolerance" value="0.05" />
	</include>

	<node pkg="statemachine_additions" type="bootUpNode"
		name="bootUpNode" output="screen">
		<param name="wait_time" value="2" />
	</node>

	<node pkg="statemachine_additions"
		type="additionsServiceProviderNode"
		name="additionsServiceProviderNode">
		<param name="update_frequency" value="20" />
		<param name="autonomy_cmd_vel_topic"
			value="$(arg autonomy_cmd_vel_topic)" />
		<param name="navigation_plugin" value="$(arg navigation_plugin)" />
	</node>

	<node pkg="move_base" type="move_base" respawn="false"
		name="move_base" output="screen">
		<rosparam
			file="$(find statemachine_additions)/config/schroedi/costmap_common.yaml"
			command="load" ns="global_costmap" />
		<rosparam
			file="$(find statemachine_additions)/config/schroedi/costmap_common.yaml"
			command="load" ns="local_costmap" />
		<rosparam
			file="$(find statemachine_additions)/config/schroedi/costmap_local.yaml"
			command="load" />
		<rosparam
			file="$(find statemachine_additions)/config/schroedi/costmap_global.yaml"
			command="load" />
		<rosparam
			file="$(find statemachine_additions)/config/schroedi/planner.yaml"
			command="load" />
		<param name="global_costmap/robot_base_frame"
			value="$(arg 
		robot_frame)" />
		<param name="local_costmap/robot_base_frame"
			value="$(arg 
		robot_frame)" />
		<remap from="/cmd_vel" to="$(arg autonomy_cmd_vel_topic)" />
	</node>

	<node pkg="move_base" type="move_base" respawn="false"
		name="move_base_reverse" output="screen">
		<rosparam
			file="$(find statemachine_additions)/config/schroedi/costmap_common.yaml"
			command="load" ns="global_costmap" />
		<rosparam
			file="$(find statemachine_additions)/config/schroedi/costmap_common.yaml"
			command="load" ns="local_costmap" />
		<rosparam
			file="$(find statemachine_additions)/config/schroedi/costmap_local.yaml"
			command="load" />
		<rosparam
			file="$(find statemachine_additions)/config/schroedi/costmap_global.yaml"
			command="load" />
		<rosparam
			file="$(find statemachine_additions)/config/schroedi/planner.yaml"
			command="load" />
		<param name="global_costmap/robot_base_frame"
			value="$(arg robot_frame)_reverse" />
		<param name="local_costmap/robot_base_frame"
			value="$(arg robot_frame)_reverse" />
		<remap from="/cmd_vel"
			to="$(arg autonomy_cmd_vel_topic)_reverse" />
		<remap from="move_base/goal" to="move_base_reverse/goal" />
		<remap from="move_base/cancel" to="move_base_reverse/cancel" />
		<remap from="move_base/feedback" to="move_base_reverse/feedback" />
		<remap from="move_base/status" to="move_base_reverse/status" />
		<remap from="move_base/result" to="move_base_reverse/result" />
	</node>

	<node pkg="tf" type="static_transform_publisher"
		name="base_footprint_reverse"
		args="0 0 0 3.1415 0 0             $(arg robot_frame)         $(arg robot_frame)_reverse		10" />


	<!-- Explore Lite -->
	<node pkg="explore_lite" type="explore" respawn="false"
		name="explore">
		<param name="robot_base_frame" value="$(arg robot_frame)" />
		<param name="costmap_topic" value="map" />
		<param name="costmap_updates_topic" value="map_updates" />
		<param name="visualize" value="true" />
		<param name="planner_frequency" value="5" />
		<param name="progress_timeout" value="3600.0" />
		<param name="potential_scale" value="3.0" />
		<param name="orientation_scale" value="0.0" />
		<param name="gain_scale" value="1.0" />
		<param name="transform_tolerance" value="0.3" />
		<param name="min_frontier_size" value="0.75" />
		<remap from="/move_base" to="/frontier_move_base" />
	</node>
</launch>